docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,286 @@
+NextGen-Core
+============
+
+This document provides a brief overview of coala's NextGen-Core.
+coala's NextGen-Core comes with the promise of resolving the dependencies
+amongst bears more efficiently (by removing the queuing mechanism) and better
+performance
+
+What is new?
+------------
+
+The following new features have been added as a part of the NextGen-Core:
+
+- Easier Interface
+- Official support for virtual files
+- Improved dependency system
+- New base bear type: ``DependencyBear``
+- Ability to modify bear dependencies at runtime
+- Custom parallelization for bears
+- Superior caching
+
+What has changed?
+-----------------
+
+- ``Global Bears`` are now called ``Project Bears`` and ``Local Bears`` are
+  now known as ``File Bears``. Both the ``ProjectBear`` and the ``FileBear``
+  classes inherit from the base class ``coalib.core.Bear``.
+- There is no need to pass ``HiddenResults`` between different bears (made
+  possible by the new dependency management system) to hide results to the
+  result callback. Only the results that were explicitly requested by passing
+  the needed bears are passed now. There is no need to pass instances of
+  ``Result`` anymore, any python object will do.
+- The former ``run`` function that was inherited by all the bears to run code
+  analysis is now replaced by the ``analyze`` function.
+
+Easier Interface
+----------------
+
+Running a coala session with the NextGen-Core is now as easy as
+``core.run(bears)``. The ``run`` method takes the following arguments ``bears``,
+``result_callback``, ``cache`` and ``executor`` (the last two are ``None`` by
+default) and initiates a coala session.
+
+* The ``bears`` argument contains the list of bears to be run for a coala
+  session.
+* The ``result_callback`` is a function that is called when the results from a
+  bear run are available.
+* The ``cache`` argument if provided enables caching and runs the session using
+  the cache provided to store the bear results.
+* The ``executor`` argument is used to provide a custom executor in which the
+  passed bears are to be run which is closed after the core is closed. If this
+  argument is not provided then ``ProcessPoolExecutor`` is used using as many
+  processes as cores available on the system.
+
```

Comment:
I think you should also mention here that bears always have to be constructed with `section` and `filedict`.
This is part of the core contract that makes the dependency system possible.

Reply:
/ the consequence is that bear-writers have to care when writing their own `__init__`. Default parameters are allowed as many as wanted, but not positional parameters (default parameters are btw discouraged, as the dependency system automatically instantiates bears, there might be unwanted side effects).